### PR TITLE
Add BibTex/APALIKE citation string converter

### PR DESCRIPTION
**Pull request by paxos**
Greetings 👋 ,
this PR aims to add support to generate BibTex and APALIKE citation strings.

It also includes a test suite (unfortunately only with a very minimal set of test files yet) and the fixtures are loosely inspired by the way https://github.com/citation-file-format/cff-converter-python is generating those strings (with a slight variation on the way how author names are generated). I used [mendeleys](https://www.mendeley.com/guides/apa-citation-guide) docs as a reference for the author name citation.

Let me know if there are any changes required. Thank you for considering this Pull Request!

related: https://github.com/citation-file-format/ruby-cff/issues/17

cc @arfon

### CODE REVIEW

**Comment by paxos** [comment]
Some tests are failing with the new fixtures (which is great!), will fix them asap

**Comment by paxos** [comment]
It seems some of the crashes are not directly related to the changes in this PR. There is a potential bug in the parsing logic for `references`:

```
Minitest::UnexpectedError: NoMethodError: undefined method `map!' for nil:NilClass
    /Users/pd/work/ruby-cff/lib/cff/util.rb:31:in `build_actor_collection!'
    /Users/pd/work/ruby-cff/lib/cff/reference.rb:295:in `block in build_model'
    /Users/pd/work/ruby-cff/lib/cff/reference.rb:294:in `each'
    /Users/pd/work/ruby-cff/lib/cff/reference.rb:294:in `build_model'
    /Users/pd/work/ruby-cff/lib/cff/reference.rb:140:in `initialize'
    /Users/pd/work/ruby-cff/lib/cff/model.rb:120:in `new'
    /Users/pd/work/ruby-cff/lib/cff/model.rb:120:in `block in build_model'
    /Users/pd/work/ruby-cff/lib/cff/model.rb:119:in `map!'
    /Users/pd/work/ruby-cff/lib/cff/model.rb:119:in `build_model'
    /Users/pd/work/ruby-cff/lib/cff/model.rb:64:in `initialize'
    /Users/pd/work/ruby-cff/lib/cff/file.rb:54:in `new'
    /Users/pd/work/ruby-cff/lib/cff/file.rb:54:in `initialize'
    /Users/pd/work/ruby-cff/lib/cff/file.rb:68:in `new'
    /Users/pd/work/ruby-cff/lib/cff/file.rb:68:in `read'
    /Users/pd/work/ruby-cff/test/cff_apa_formatter_test.rb:11:in `block (2 levels) in <class:CFFApaFormatterTest>'
```

Planning to take a look on those, most likely opening a separate PR for this.

File: lib/cff/file.rb [hunk 1]
```diff
@@ -65,7 +65,7 @@ def self.read(file)
       content = ::File.read(file)
       comment = File.parse_comment(content)
 
-      new(YAML.safe_load(content, [Date]), comment)
+      new(YAML.safe_load(content, [Date, Time]), comment)
```
Comment (paxos) [inline lib/cff/file.rb:68]: This was needed for some of the fixtures; not sure if this is the best solution tho 🤔 thoughts?

File: lib/cff/version.rb [hunk 2]
```diff
@@ -17,6 +17,6 @@
 ##
 module CFF
   # :nodoc:
-  VERSION = '0.5.0'
+  VERSION = '0.6.0'
```
Comment (paxos) [inline lib/cff/version.rb:20]: Bump the version; feel free to edit :)

File: lib/cff/reference.rb [hunk 3]
```diff
@@ -148,7 +148,7 @@ def initialize(param, *more)
       [
         'authors', 'contact', 'editors', 'editors-series', 'keywords',
         'patent-states', 'recipients', 'senders', 'translators'
-      ].each { |field| @fields[field] = [] if @fields[field].empty? }
+      ].each { |field| @fields[field] = [] if !@fields[field].nil? && @fields[field].empty? }
```
Comment (paxos) [inline lib/cff/reference.rb:151]: Bugfix

File: lib/cff/reference.rb [hunk 4]
```diff
@@ -292,7 +292,7 @@ def build_model(fields) # :nodoc:
         'authors', 'contact', 'editors', 'editors-series', 'recipients',
         'senders', 'translators'
       ].each do |field|
-        build_actor_collection!(fields[field])
+        build_actor_collection!(fields[field]) if fields.include?(field)
```
Comment (paxos) [inline lib/cff/reference.rb:295]: Bugfix

**Comment by paxos** [comment]
All tests are now ✅ and this PR is ready for review!

**Comment by arfon** [comment]
> All tests are now ✅ and this PR is ready for review!

:zap: thanks @paxos. @hainesr – would love your feedback on these changes when you get a moment.

File: lib/cff/version.rb (hunk 2 above)
Comment (hainesr) [reply lib/cff/version.rb:20]: 👍

File: lib/cff/file.rb (hunk 1 above)
Comment (hainesr) [reply lib/cff/file.rb:68]: I'm surprised it's needed as I thought CFF only used dates and not full timestamps. Given the discussion over at citation-file-format/citation-file-format#202 we may not need either `Date` or `Time` for version 2.0?

File: lib/cff/reference.rb (hunk 3 above)
Comment (hainesr) [reply lib/cff/reference.rb:151]: Good catch, thanks

File: lib/cff/reference.rb (hunk 4 above)
Comment (hainesr) [reply lib/cff/reference.rb:295]: catch, thanks

**Comment by hainesr** [comment]
Thanks for this @paxos and @arfon!

I think I made a few changes to the master branch since you started work on this. I'm going to pull this locally and rebase and see where we get to...

**Comment by paxos** [comment]
Thank you @hainesr for taking a look!

I resolved the merge conflict and merged in current master. Let me know if there is something I can do to help reviewing this :)

**Comment by paxos** [comment]
This is what I used to generate the test fixtures: https://github.com/paxos/ruby-cff-fixtures-generator

I had to adjust some of the output files, as not everything was converted correctly.

**Comment by hainesr** [comment]
Thanks @paxos. My merge commit was the same as yours, so I just fast-forwarded master and this is now merged!

File: lib/cff/file.rb (hunk 1 above)
Comment (paxos) [reply lib/cff/file.rb]: Looks good! In case we do, the tests should catch that!

**Pull request closed without merging**